Unable to transfer file to pod if /tmp of pod is read-only

Since changes from https://github.com/fabric8io/kubernetes-client/pull/4968/commits/86e740fbb84e073f950a34c891ee95ffa619fd92 have been implemented, we have an error when transferring a file to a pod that has /tmp read-only.

Before these changes (fix #4910), it was transferring the file to a directory that is allowed to write, but with the new way to transfer files, it seems it must have write access to /tmp.

To answer the obvious question you will ask as to is it possible to have /tmp writable in our pod, the answer is no, it is impossible.

What we suspect: upload function in PodUpload.java (https://github.com/fabric8io/kubernetes-client/blob/v6.8.1/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/uploadable/PodUpload.java#L135)

Version with the issue: 6.8.1

**Question**:
Is there a way (or an option) to have the file being copied directly in the intended directory and not have it copied to /tmp first?
(We need to upgrade version of kubernetes-client, and this is a blocker)

Error message:

Caused by: io.fabric8.kubernetes.client.KubernetesClientException: sh: /tmp/fabric8-a8874cdd-eb99-4f4c-a022-20150120503f.tar: Read-only file system

        at io.fabric8.kubernetes.client.dsl.internal.ExecWebSocketListener.onMessage(ExecWebSocketListener.java:319)
        at io.fabric8.kubernetes.client.okhttp.OkHttpWebSocketImpl$1.onMessage(OkHttpWebSocketImpl.java:110)
        at okhttp3.internal.ws.RealWebSocket.onReadMessage(RealWebSocket.kt:338)
        at okhttp3.internal.ws.WebSocketReader.readMessageFrame(WebSocketReader.kt:247)
        at okhttp3.internal.ws.WebSocketReader.processNextFrame(WebSocketReader.kt:106)
        at okhttp3.internal.ws.RealWebSocket.loopReader(RealWebSocket.kt:293)
        at okhttp3.internal.ws.RealWebSocket$connect$1.onResponse(RealWebSocket.kt:195)
        at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:519)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
        Suppressed: java.lang.Throwable: waiting here
                at io.fabric8.kubernetes.client.utils.Utils.waitUntilReady(Utils.java:174)
                at io.fabric8.kubernetes.client.dsl.internal.uploadable.PodUpload.upload(PodUpload.java:101)
                at io.fabric8.kubernetes.client.dsl.internal.uploadable.PodUpload.uploadTar(PodUpload.java:141)
                at io.fabric8.kubernetes.client.dsl.internal.uploadable.PodUpload.upload(PodUpload.java:55)
                at io.fabric8.kubernetes.client.dsl.internal.core.v1.PodOperationsImpl.lambda$upload$4(PodOperationsImpl.java:427)
                at io.fabric8.kubernetes.client.utils.internal.OptionalDependencyWrapper.wrapRunWithOptionalDependency(OptionalDependencyWrapper.java:39)
                at io.fabric8.kubernetes.client.dsl.internal.core.v1.PodOperationsImpl.upload(PodOperationsImpl.java:425)
…

Thank you

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@
 * Fix #5580: [java-generator] Correctly handle defaults for IntOrString types
 * Fix #5584: Fix CRD generation when EnumMap is used
 * Fix #5626: Prevent memory accumulation from informer usage
+* Fix #5527: Unable to transfer file to pod if `/tmp` is read-only
 
 #### Improvements
 * Fix #5429: moved crd generator annotations to generator-annotations instead of crd-generator-api. Using generator-annotations introduces no transitive dependencies.

--- a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/uploadable/PodUpload.java
+++ b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/uploadable/PodUpload.java
@@ -52,10 +52,10 @@ public class PodUpload {
       throws IOException {
 
     if (Utils.isNotNullOrEmpty(operation.getContext().getFile()) && pathToUpload.toFile().isFile()) {
-      return uploadTar(operation, getDirectoryFromFile(operation),
+      return uploadTar(operation, getDirectoryFromFile(operation.getContext().getFile()),
           tar -> addFileToTar(new File(operation.getContext().getFile()).getName(), pathToUpload.toFile(), tar));
     } else if (Utils.isNotNullOrEmpty(operation.getContext().getDir()) && pathToUpload.toFile().isDirectory()) {
-      return uploadTar(operation, operation.getContext().getDir(), tar -> {
+      return uploadTar(operation, ensureEndsWithSlash(operation.getContext().getDir()), tar -> {
         for (File file : pathToUpload.toFile().listFiles()) {
           addFileToTar(file.getName(), file, tar);
         }
@@ -64,10 +64,9 @@ public class PodUpload {
     throw new IllegalArgumentException("Provided arguments are not valid (file, directory, path)");
   }
 
-  private static String getDirectoryFromFile(PodOperationsImpl operation) {
-    String file = operation.getContext().getFile();
+  private static String getDirectoryFromFile(String file) {
     String directoryTrimmedFromFilePath = file.substring(0, file.lastIndexOf('/'));
-    return directoryTrimmedFromFilePath.isEmpty() ? "/" : directoryTrimmedFromFilePath;
+    return ensureEndsWithSlash(directoryTrimmedFromFilePath.isEmpty() ? "/" : directoryTrimmedFromFilePath);
   }
 
   private interface UploadProcessor<T extends OutputStream> {
@@ -136,7 +135,7 @@ public class PodUpload {
       UploadProcessor<TarArchiveOutputStream> processor)
       throws IOException {
 
-    String fileName = String.format("/tmp/fabric8-%s.tar", UUID.randomUUID());
+    String fileName = String.format("%sfabric8-%s.tar", directory, UUID.randomUUID());
 
     boolean uploaded = upload(operation, fileName, os -> {
       try (final TarArchiveOutputStream tar = new TarArchiveOutputStream(os)) {
@@ -188,10 +187,12 @@ public class PodUpload {
   }
 
   static String createExecCommandForUpload(String file) {
-    String directoryTrimmedFromFilePath = file.substring(0, file.lastIndexOf('/'));
-    final String directory = directoryTrimmedFromFilePath.isEmpty() ? "/" : directoryTrimmedFromFilePath;
     return String.format(
-        "mkdir -p %s && cat - > %s", shellQuote(directory), shellQuote(file));
+        "mkdir -p %s && cat - > %s", shellQuote(getDirectoryFromFile(file)), shellQuote(file));
+  }
+
+  private static String ensureEndsWithSlash(String path) {
+    return path.endsWith("/") ? path : (path + "/");
   }
 
 }

--- a/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/behavior/UploadTest.java
+++ b/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/behavior/UploadTest.java
@@ -135,7 +135,7 @@ class UploadTest {
         .extracting(StandardWebSocketBuilder::asHttpRequest)
         .extracting(StandardHttpRequest::uri)
         .extracting(URI::getQuery).asString()
-        .contains("command=rm /tmp/fabric8-");
+        .contains("command=rm /target/fabric8-");
   }
 
   @Nested
@@ -213,7 +213,7 @@ class UploadTest {
             .extracting(StandardWebSocketBuilder::asHttpRequest)
             .extracting(StandardHttpRequest::uri)
             .extracting(URI::getQuery).asString()
-            .contains("command=mkdir -p '/tmp' && cat - > '/tmp/fabric8-");
+            .contains("command=mkdir -p '/target/' && cat - > '/target/fabric8-");
       }
 
       @Test
@@ -229,7 +229,7 @@ class UploadTest {
             .extracting(StandardWebSocketBuilder::asHttpRequest)
             .extracting(StandardHttpRequest::uri)
             .extracting(URI::getQuery).asString()
-            .contains("command=wc -c < '/tmp/fabric8-");
+            .contains("command=wc -c < '/target/fabric8-");
       }
 
       @Test
@@ -246,7 +246,7 @@ class UploadTest {
             .extracting(StandardHttpRequest::uri)
             .extracting(URI::getQuery).asString()
             .matches(
-                ".+command=mkdir -p '/target-dir'; tar -C '/target-dir' -xmf /tmp/fabric8-.+\\.tar; e=\\$\\?; rm /tmp/fabric8-.+");
+                ".+command=mkdir -p '/target-dir/'; tar -C '/target-dir/' -xmf /target-dir/fabric8-.+\\.tar; e=\\$\\?; rm /target-dir/fabric8-.+");
       }
 
       @Nested
@@ -272,7 +272,7 @@ class UploadTest {
           final byte[] tarBytes = new byte[sendCaptor.getValue().remaining() - 1];
           System.arraycopy(sendCaptor.getValue().array(), 1, tarBytes, 0, tarBytes.length);
           final TarArchiveInputStream tar = new TarArchiveInputStream(new ByteArrayInputStream(tarBytes));
-          assertThat(tar.getNextTarEntry())
+          assertThat(tar.getNextEntry())
               .hasFieldOrPropertyWithValue("name", "file-name.txt")
               .hasFieldOrPropertyWithValue("size", toUpload.toFile().length());
         }
@@ -292,7 +292,7 @@ class UploadTest {
           final byte[] tarBytes = new byte[sendCaptor.getValue().remaining() - 1];
           System.arraycopy(sendCaptor.getValue().array(), 1, tarBytes, 0, tarBytes.length);
           final TarArchiveInputStream tar = new TarArchiveInputStream(new ByteArrayInputStream(tarBytes));
-          assertThat(tar.getNextTarEntry())
+          assertThat(tar.getNextEntry())
               .hasFieldOrPropertyWithValue("name", longFileName);
         }
 
@@ -310,7 +310,7 @@ class UploadTest {
           final byte[] tarBytes = new byte[sendCaptor.getValue().remaining() - 1];
           System.arraycopy(sendCaptor.getValue().array(), 1, tarBytes, 0, tarBytes.length);
           final TarArchiveInputStream tar = new TarArchiveInputStream(new ByteArrayInputStream(tarBytes));
-          assertThat(tar.getNextTarEntry())
+          assertThat(tar.getNextEntry())
               .hasFieldOrPropertyWithValue("name", "file-name.txt")
               .hasFieldOrPropertyWithValue("lastModifiedTime", FileTime.fromMillis(9999999999999L));
         }
@@ -351,7 +351,7 @@ class UploadTest {
             .extracting(StandardWebSocketBuilder::asHttpRequest)
             .extracting(StandardHttpRequest::uri)
             .extracting(URI::getQuery).asString()
-            .contains("command=mkdir -p '/tmp' && cat - > '/tmp/fabric8-");
+            .contains("command=mkdir -p '/target/location/' && cat - > '/target/location/fabric8-");
       }
 
       @Test
@@ -367,7 +367,7 @@ class UploadTest {
             .extracting(StandardWebSocketBuilder::asHttpRequest)
             .extracting(StandardHttpRequest::uri)
             .extracting(URI::getQuery).asString()
-            .contains("command=wc -c < '/tmp/fabric8-");
+            .contains("command=wc -c < '/target/location/fabric8-");
       }
 
       @Test
@@ -384,7 +384,7 @@ class UploadTest {
             .extracting(StandardHttpRequest::uri)
             .extracting(URI::getQuery).asString()
             .matches(
-                ".+command=mkdir -p '/target-dir/location'; tar -C '/target-dir/location' -xmf /tmp/fabric8-.+\\.tar; e=\\$\\?; rm /tmp/fabric8-.+");
+                ".+command=mkdir -p '/target-dir/location/'; tar -C '/target-dir/location/' -xmf /target-dir/location/fabric8-.+\\.tar; e=\\$\\?; rm /target-dir/location/fabric8-.+");
       }
 
     }
@@ -423,7 +423,7 @@ class UploadTest {
             .extracting(StandardWebSocketBuilder::asHttpRequest)
             .extracting(StandardHttpRequest::uri)
             .extracting(URI::getQuery).asString()
-            .contains("command=mkdir -p '/target' && cat - > '/target/location");
+            .contains("command=mkdir -p '/target/' && cat - > '/target/location");
       }
 
       @Test

--- a/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/dsl/internal/uploadable/PodUploadTest.java
+++ b/kubernetes-client/src/test/java/io/fabric8/kubernetes/client/dsl/internal/uploadable/PodUploadTest.java
@@ -57,7 +57,7 @@ class PodUploadTest {
       // When
       String result = PodUpload.createExecCommandForUpload("/tmp/foo/cp.log");
       // Then
-      assertThat(result).isEqualTo("mkdir -p '/tmp/foo' && cat - > '/tmp/foo/cp.log'");
+      assertThat(result).isEqualTo("mkdir -p '/tmp/foo/' && cat - > '/tmp/foo/cp.log'");
     }
 
     //
@@ -66,7 +66,7 @@ class PodUploadTest {
       // When
       String result = PodUpload.createExecCommandForUpload("/tmp/fo'o/cp.log");
       // Then
-      assertThat(result).isEqualTo("mkdir -p '/tmp/fo\'\\'\'o' && cat - > '/tmp/fo\'\\'\'o/cp.log'");
+      assertThat(result).isEqualTo("mkdir -p '/tmp/fo\'\\'\'o/' && cat - > '/tmp/fo\'\\'\'o/cp.log'");
     }
 
     @Test
@@ -74,7 +74,7 @@ class PodUploadTest {
       // When
       String result = PodUpload.createExecCommandForUpload("/tmp/f'o'o/c'p.log");
       // Then
-      assertThat(result).isEqualTo("mkdir -p '/tmp/f\'\\'\'o\'\\'\'o' && cat - > '/tmp/f\'\\'\'o\'\\'\'o/c\'\\'\'p.log'");
+      assertThat(result).isEqualTo("mkdir -p '/tmp/f\'\\'\'o\'\\'\'o/' && cat - > '/tmp/f\'\\'\'o\'\\'\'o/c\'\\'\'p.log'");
     }
   }
 

